// auth0-lock/lib/ui/button/auth_button.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AuthButton = function AuthButton(props) {
  var disabled = props.disabled;
  var isBig = props.isBig;
  var label = props.label;
  var onClick = props.onClick;
  var strategy = props.strategy;


  var className = "auth0-lock-social-button";
  if (isBig) className += " auth0-lock-social-big-button";

  return _react2.default.createElement(
    "button",
    {
      className: className,
      "data-provider": strategy,
      disabled: disabled,
      onClick: onClick,
      type: "button"
    },
    _react2.default.createElement("div", { className: "auth0-lock-social-button-icon" }),
    _react2.default.createElement(
      "div",
      { className: "auth0-lock-social-button-text" },
      label
    )
  );
};

AuthButton.propTypes = {
  disabled: _react2.default.PropTypes.bool.isRequired,
  isBig: _react2.default.PropTypes.bool.isRequired,
  label: _react2.default.PropTypes.string.isRequired,
  onClick: _react2.default.PropTypes.func.isRequired,
  strategy: _react2.default.PropTypes.string.isRequired
};

AuthButton.defaultProps = {
  disabled: false,
  isBig: true
};

exports.default = AuthButton;
